File: src/endpoints/comment/get_index.py
from . import __types__ as types
from src import output
from src.libs.db import DB, Table
from src.modules.verify import checking_token

async def get_index(params: dict = {}, req = None, _db: DB = None, _check_token = True):

    # set values
    result = None
    db = _db if _db else DB().connect()

    try:
        # set params
        params = types.GetIndex(**params)

        # checking token
        if _check_token: checking_token(req, db)

        # set fields
        fields = params.fields.split(',') if params.fields else None

        # set where
        where = []
        if params.module:
            where.append(f'and module LIKE "{params.module}"')
        if params.module_srl:
            where.append(f'and module_srl = {params.module_srl}')
        if params.content:
            where.append(f'and content LIKE "%{params.content}%"')

        # set values
        values = {}

        # get total
        total = db.get_count(
            table_name = Table.COMMENT.value,
            where = where,
            values = values,
        )
        if not (total > 0): raise Exception('No data', 204)

        # get data
        index = db.get_items(
            table_name = Table.COMMENT.value,
            fields = fields,
            where = where,
            limit = {
                'size': params.size,
                'page': params.page,
            },
            order = {
                'order': params.order,
                'sort': params.sort,
            },
            values = values,
            unlimited = params.unlimited,
        )

        # set result
        result = output.success({
            'message': 'Complete get comment index.',
            'data': {
                'total': total,
                'index': index,
            },
        }, _req=req)
    except Exception as e:
        result = output.exc(e, _req=req)
    finally:
        if not _db and db: db.disconnect()
        return result

File: src/routes/comment/get-index.ts
import DB, { db } from '@/classes/DB'
import ServiceError from '@/classes/ServiceError'
import Service from '@/classes/Service'
import type { CommentModel } from './__model'

type GetIndexParams = {
  query: CommentModel['getIndexQuery']
  service: Service
}

export default async function getItem({ query, service }: GetIndexParams)
{
  try
  {
    // set assets
    let _where: string[] = []
    let _values: ZZ = {}
    const _field = query.field ? query.field.split(',') : ''

    // set base params
    if (query.module)
    {
      _where.push(`AND module LIKE $module`)
      _values['$module'] = query.module
    }
    if (query.module && query.module_srl)
    {
      _where.push(`AND module_srl = ${query.module_srl}`)
    }
    if (query.q)
    {
      _where.push(`AND content LIKE '%' || $content || '%'`)
      _values['$content'] = query.q
    }

    // get count
    const count = db.getCount({
      table: DB.TABLE.COMMENT,
      where: _where,
      values: _values,
    })
    if (count.data <= 0) throw new ServiceError('No data', { status: 204 })

    // get index
    let index = db.getIndex({
      table: DB.TABLE.COMMENT,
      field: _field,
      where: _where,
      order: Boolean(query.order || query.sort) ? query.order : 'srl',
      sort: Boolean(query.order || query.sort) ? query.sort : 'desc',
      page: query.page ?? 1,
      size: query.size ?? service.preference['comment.index.size'],
      values: _values,
    })

    return {
      total: count.data,
      index: index.data,
    }
  }
  catch (_e: any)
  {
    throw new ServiceError('Failed to get Comment index.', {
      status: _e.status,
      text: _e.message,
      cause: _e,
    })
  }
}
